### DISCUSSION TITLE
Lesson 4, Concept 12. Key pairs used by groups of people - how is the private key kept secret?

### DISCUSSION THREAD

**Discussion by JohannesStutz** [OpenMined/courses#214] [Course: Our Privacy Opportunity] [answered]
Hi! Regarding Concept 12, Active Security & Zero Knowledge Proofs. Key pairs can be used anonymously by groups of people. I understand that every member of this group has to have the private key, is this correct? In larger groups, how is it possible to trust every member not to give this key to an unauthorized person? Or even if there is no bad intention, the risk of theft must be much higher when many people are involved?

Thank you for answering basic questions :)

**Comment by JohannesStutz** [comment, marked as answer]
Or is it like this: every member of the group has their own private key. But the associated public key is known to belong to the group, thus allowing proof of membership?
But then this would make the whistleblower example impossible... Because the organization could link the key to an individual, right?
I'm somewhat confused :D

**Reply by Nilanshrajput** [reply]
@JohannesStutz  Every member of the group have their own key pairs, now for the second point now the organisation could not link to an individual as firstly a public/private key pair doesn’t have to be tied to an exact identity you could remain anonymous no information about you is revealed everyone only knows that some random person X has a signature and X doesn't have a name, gender or any personal information associated, you can refer to this great summary from @nahuakang  https://www.nahua.dev/series/privacy_ai_course_notes/course-1-lesson-4-summary/ !

**Reply by leriomaggio** [reply]
@JohannesStutz this is very interesting question and I feel I can only complement the quite comprehensive answers you got already by offering a slightly different perspective - also related to the _group membership_ example.

The whole concept of key-based encryption methods relies on using a pair of cryptographic keys, namely **private** (`prv_key`) and **public** keys (`pub_key`)  to be respectively used to _sign_ and to _verify_ the information in an information flow. This strategy of encryption is also known as _Asymmetric Encryption_ (here is an [article](https://searchsecurity.techtarget.com/definition/asymmetric-cryptography) if you would like to know more).

With particular reference to the _Input Verification_ strategy (also the core of the concept in the course you are mentioning here) this method is able to verify that the information signed is indeed authentic, and the very concept of _authenticity_ is the key to understand benefits and limitations of this approach. 

In fact, **if** the signature produced by `prv_key` _matches_ the verification produced by the `pub_key` you have guarantee at the same time that (A) the information has not been altered within the information flow, and (B) the sender of indeed who is claiming to be (i.e. the _key owner_).

This is indeed why in the video Andrew mentioned that this method provides both Input Privacy _and_ Input Verification at the same time. On the other hand, this method has not an extensive applicability, being limited to "direct" information flows (i.e. _one sender, one receiver_ scenarios).

Now, the whole approach is crucially based on **one** very simple key concept: 
> The Private Key has to be kept - and remain - private! 

As soon as thist requirement isn't guaranteed anymore, the whole mechanisms is flawed. This may happen either because the key has been stolen (i.e. malicious intentions), or because the key has been borrowed to someone else by the original owner.

**Note** Nothing is said about the `pub_key`! Well, that's because the public key has to be _public_ by definition otherwise _verification_ wouldn't be possible, and also because the `pub_key` alone is completely useless. 

And this goes to the second important feature of this approach: a Public/Private Key pair is not _tight_ to a specific person but only to what @iamtrask referred to as the "Anonymous Identity". Considering the example of the group you were mentioning, a single key pair does not identify a single individual (as in _human being_) but only a specific group member.
The information disclosed to the group by the single members depend on the group policy (e.g. "members are identified by real names or alias"), and it varies depending on the context. 

This is an important concept to keep in mind, because this would resolve your concern about "the group knows who I am": no, group owners and/or group members only know who you are "within the group", as in `@leriomaggio` and not `@JohannesStutz`.  

Now, to finally go to your last point, as in "proof of membership". Using this strategy, the whole approach could be implemented in this way:

The group has access to the list of "allowed" `pub_keys` of members - and that's by definition, as this is the only disclosable information they can access from members, and also the only piece of information needed to allow verification.
 
Whenever the "input verification" is required in an information flow, the proof of membership can be guaranteed to the extent of verifying with `100%` certainty that a specific message has been sent by the sender you were expecting.

So, if you were expecting the sender to be `jolly_sunday365` which is also a member of the group, and so you have access to their `pub_key`, this method guarantees that **only** the other half key `prv_key` of the pair could have been used to sign the received  message. If that's not the case, the verification fails.

I hope this clarifies a little bit about the method :) If you're happy with our answers, please feel free to mark the question as "Answered"

**Reply by JohannesStutz** [reply]
Thank you SO much for this great explanation, Valerio! This makes much more sense to me now. I think I 99% got it 😄 

I will have to read a bit more about public-key-infrastructure, how do members get their public key without the organization knowing their exact identity.

Sorry that I had to mark my own comment as the answer, since your explanations were comments to this instead of the original question.

Thank you guys again!

**Comment by znreza** [comment]
@JohannesStutz  adding to what @Nilanshrajput  mentioned,  since every member have their unique private-public key pairs, they don't have to share their private keys with anyone. Only the public key will be know by the other members. Also, if someone wants to prevent an entity to link him/her identity to the key he/she shared, he/she can use multiple key pairs to communicate with different entites an preserve their anonymity.
